exporters/otlp/otlphttp/driver.go

```diff
@@ -0,0 +1,269 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otlphttp
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"path"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp"
+	colmetricspb "go.opentelemetry.io/otel/exporters/otlp/internal/opentelemetry-proto-gen/collector/metrics/v1"
+	coltracepb "go.opentelemetry.io/otel/exporters/otlp/internal/opentelemetry-proto-gen/collector/trace/v1"
+	"go.opentelemetry.io/otel/exporters/otlp/internal/transform"
+	metricsdk "go.opentelemetry.io/otel/sdk/export/metric"
+	tracesdk "go.opentelemetry.io/otel/sdk/export/trace"
+)
+
+const contentType = "application/x-protobuf"
+
+var defaultTransportClone *http.Transport
+
+func init() {
+	// Clone default transport early, possibly before it's
+	// used/modified.
+	realTransport := http.DefaultTransport.(*http.Transport)
```

Comment:
Is it worth having a more explicit error message if this cast fails (because this package was initialized after something else that replaced `http.DefaultTransport`?
```suggestion
	realTransport, ok := http.DefaultTransport.(*http.Transport)
	if !ok {
		panic(fmt.Sprintf("http.DefaultTransport has been replaced with a %T due to package initialization order. The OTLP HTTP transporter requires it to be a *http.Transport.", http.DefaultTransport))
	}
```

Reply:
Is our concern here that someone may have changed the `http.DefaultTransport` from its stdlib-provided configuration or that it might be an `http.RoundTripper` that is not implemented through `*http.Transport`?  In either case, I don't think that we can `panic()` if our expectations do not hold.  We should perhaps instead prepare a new `http.Transport` with an expected configuration.

Reply:
If we can't panic, then yeah, the only thing we can do is our version of http.Transport instance. It's going to be a PITA to keep it up to date with golang's `DefaultTransport` though…

Reply:
Perhaps we can emit a warning at this point if we detect that things are out of sync.

Reply:
I suppose that would be a warning during build, but for now I'm not implementing it.